crates/shim-sgx/src/handler/mod.rs

```diff
@@ -996,4 +1144,117 @@ impl<'a> Handler<'a> {
+
+        debugln!(self, "[{tid}] handle_page_fault: {addr:#x} {error_code:?}");
+
+        // check for non-valid flags
+        if !error_code
+            .intersection(
+                PageFaultErrorCode::PROTECTION_VIOLATION
+                    | PageFaultErrorCode::PROTECTION_KEY
+                    | PageFaultErrorCode::SGX,
+            )
+            .is_empty()
+        {
+            debugln!(
+                self,
+                "[{tid}] handle_page_fault: intersection {:#?}",
+                error_code.difference(
+                    PageFaultErrorCode::PROTECTION_VIOLATION
+                        | PageFaultErrorCode::PROTECTION_KEY
+                        | PageFaultErrorCode::SGX,
+                )
+            );
+
+            return Err(());
+        }
+
+        // check for must have flags
+        if !error_code.contains(PageFaultErrorCode::USER_MODE) {
+            return Err(());
+        }
+
+        let mut heap = HEAP.write();
+
+        let addr = Address::new(addr);
+        let length = Offset::from_items(1);
+
+        let access = match heap.contains(addr, length) {
+            None => {
+                debugln!(
+                    self,
+                    "[{tid}] handle_page_fault: access empty {:#?}",
+                    Region::new(addr, addr + length),
+                );
+                panic!();
+            }
+            Some(access) => {
+                // check for real page fault
+                if !access.contains(Access::READ)
+                    || (error_code.contains(PageFaultErrorCode::CAUSED_BY_WRITE)
+                        && !access.contains(Access::WRITE))
+                    || (error_code.contains(PageFaultErrorCode::INSTRUCTION_FETCH)
+                        && !access.contains(Access::READ | Access::EXECUTE))
+                {
+                    debugln!(
+                        self,
+                        "[{tid}] handle_page_fault: real page fault previous access {:#?}",
+                        Record {
+                            region: Region::new(addr, addr + length),
+                            access
+                        }
+                    );
+                    panic!();
+                }
+                access
+            }
+        };
+
+        if heap.mmap(Some(addr), length, access | Access::MMAPPED) != Some(addr) {
+            debugln!(
+                self,
+                "[{tid}] handle_page_fault: MMAPPED heap is inconsistent"
+            );
+            panic!();
+        }
+
+        if let Err(e) = self.mmap_host(
+            NonNull::new(addr.raw() as *mut _).unwrap(),
+            length.bytes(),
+            PROT_READ | PROT_WRITE,
+        ) {
+            debugln!(
+                self,
+                "[{tid}] handle_page_fault: ERROR mmap_host() = {e:#?}"
+            );
+            panic!();
+        }
+
+        self.mmap_guest(addr, length, flags_from_access(access));
```

Comment:
Is this operation infallible?

Reply:
It unwraps internally and should really panic, because it means something is really wrong (maybe malicious host).